extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaRecorder.java

```diff
@@ -57,9 +58,19 @@ private static void shortenNodeName(TransactionManagerConfiguration transactions
         String originalNodeName = transactions.nodeName;
         log.warnf("Node name \"%s\" is longer than 28 bytes, shortening it by using %s.", originalNodeName,
                 HASH_ALGORITHM_FOR_SHORTENING);
-        final byte[] nodeNameAsBytes = originalNodeName.getBytes();
+        final byte[] nodeNameAsBytes = originalNodeName.getBytes(StandardCharsets.UTF_8);
         MessageDigest messageDigest224 = MessageDigest.getInstance(HASH_ALGORITHM_FOR_SHORTENING);
-        transactions.nodeName = new String(messageDigest224.digest(nodeNameAsBytes), StandardCharsets.UTF_8);
```

Comment:
Thanks @yrodiere for your analysis. I would like to clarify some points (I hope we all agree about them):
- only Strings longer then 28 bytes are shorten by the following process.
- the original string has an arbitrary length ( >28 bytes), then it is hashed and it's output is 28 bytes (SHA-224). This is the only hash used in this process.
- Base64 encoding is done after the byte array is hashed.
- the base64 encoded byte array has an arbitrary length and so it is truncate so that it is a 28 bytes array 
- Using the Base64 encoding for the byte array the resulting Strings are printable, are utf-8 valid and can be easily encoded to UTF-8 without changing the length

Reply:
Shouldn't this also be documented in the JavaDoc of https://github.com/quarkusio/quarkus/blob/2d4b470b10091632cee1bea4a44380a85bc71baa/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java#L35 ?

Reply:
Btw. I heard about different numbers that are actually known limits.  28 for Narayana itself and 23 for usage under Kubernetes. Would it make sense to make the actual algorithm more configurable. E.g. a number property with valid range from 1 to 28 and 28 as default. Maybe even replace `TransactionManagerConfiguration.shortenNodeNameIfNecessary` and let the default of the new property be 0 to signal no shortening at all.

Reply:
I will update the JavaDoc, thanks @graben . Concerning having a parameter to configure the length of the string I think it might be risky because the hash collision would be directly impacted when to string length is changed. So  we would need to make sure that the users know exactly about the impact of that configuration's change. However we need to remember that only string with byte array's length > 28 are shorten, other strings are not. So when the users have a strict limitation about the string length (e.g. for Kubernetes 23 as you said)  they can pass 23 bytes string and that string will not be changed.

Reply:
Good point @graben. Yes, it does. But I would say that the collision avoidance with 28 bytes is still acceptable (see the stress test) IMO, but it probably would not with a random number of bytes between 1 to 27. So again I am OK to make that number configurable but I would recommend anyway to shorten the hash with the longest string allowed (which is 28).

Reply:
IMHO, the question is if 28 is the maximum allowed by Narayana, why offer an optional  property to shorten values that are obviously wrong at all? Shouldn't Quarkus disallow values larger than 28 and force users to shorten themselves? No potential problems with collisions anymore! I know it's a bit of a radical solution, but it feels odd that users can define invalid properties and the software should solve it with a workaround with potential drawbacks. Just my 2 cents.

Reply:
Thanks @graben for your comment here. I am not sure I am following you here. You first were talking about adding an extra parameter for the length of the node name, now you are talking about removing this 'Generate right-length node name' feature (the one discussed in the issue https://github.com/quarkusio/quarkus/issues/30491 ). I might be wrong here but this PR's goal is to patch the existing implementation, not to change it. Could I maybe ask you to discuss this in a different issue or on Zulip where it is easier to get the community's feedback about this topic?
Getting back to your comment I would like to underline that the default behaviour is the one that you described: in fact the 'shortenNodeNameIfNecessary' variable is set to false by default (see https://github.com/quarkusio/quarkus/blob/main/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java#L34), the original reasons to implement this feature are discussed in the issue https://github.com/quarkusio/quarkus/issues/30491.

Reply:
Keep it as is, it just came into my mind, while following the process involving from the very first beginning. ;-)

Reply:
@graben and @marcosgopen I 'm not asking for changes here, but for completeness/clarification the 28 byte limit is needed only for XA, I think, and the 23 byte limit, so far, is only for kubernetes. But Narayana supports a number of non-XA transaction models and for most of those the size limit is 64 bytes. The 23 byte limit was introduced because the Wildfly k8s integration relies on Stateful sets and a Stateful id is 63 bytes and k8s sets the final 5 characters randomly (so we need 28 bytes minus 5 bytes for the maximum length of the nodeIdentifier - I forget why that forces it to be reduced, I'd have to have another think about why the original engineers noticed that as as being an issue).

Reply:
This is a good question @mmusgrov . As I wrote to Benjamin this PR patches https://github.com/quarkusio/quarkus/pull/36752 . If we want to be able to set the node name size configurable I would suggest to open a new PR or rethink this PR. Keep in mind that allowing the node name size to be configurable might increase the risk of name collision.

Reply:
@marcosgopen I wasn't asking a question, I was only "tying up loose ends" and providing clarity on why we enforce different size limits depending upon the particular transaction model and/or environment - i.e. it was a comment designed to contribute to the knowledge base, I wasn't asking for changes.

Reply:
Hi @mmusgrov, just a small question according to the 28 `byte` limit for XA. As far as I understand the code of the [`XID`](https://github.com/jbosstm/narayana/blob/0be17bd96db646b47b93b69ee6855eed8a39aeda/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/xa/XID.java#L28) implementation, it needs `gtrid` to be less than 64. In [`XATxConverter`](https://github.com/jbosstm/narayana/blob/main/ArjunaJTA/jta/classes/com/arjuna/ats/jta/xa/XATxConverter.java) `gtrid` is computed by concatenating `UID` as `byte` and `nodeIdentifier` as `bytes` both limited to 28 `byte`. That's a maximum of 56 < 64. Why does Narayana reserve a free space of 8 `byte`s length? Maybe I do miss some other point in the code.

Reply:
Hi @mmusgrov, you might missed my question. :-)

Reply:
Hi @graben , thanks for your question. The answer is that "in reality this must be less that 28 to fit in the bqual with the EIS name and Uid." see https://issues.redhat.com/browse/JBTM-1651

Reply:
@yrodiere Do you have more questions about this or you are OK to resolve this thread?

Reply:
I'm fine with resolving this thread, I was merely pointing out that it was important to merge the solution as-is for more reasons than just https://github.com/quarkusio/quarkus/issues/40256.
The patch was already good as far as I'm concerned :)